### PR TITLE
ntriples and turtle parser and serialised fixes

### PR DESCRIPTION
**Pull request by zozlak**
A bunch of changes to the ntriples and turtle serializer and parser:

* Ntriples serializer doesn't encode unicode characters any more (as they are valid and escaping them takes a lot of time and involves a lot of strings copying).
* Ntriples serializer applies different escaping rules to IRIs and literals (according to https://www.w3.org/TR/n-triples/#n-triples-grammar)
* Ntriples parser allows comments at the end of a line (as defined in https://www.w3.org/TR/n-triples/#n-triples-grammar)
* Turtle serializer simply uses Ntriples escape routine for escaping literals (which is perfectly valid and reduces the codebase)
* Turtle parser splits input into an array with `preg_split('//u', $data, null, PREG_SPLIT_NO_EMPTY);` which allows performant character position-based access to the parsed input. The previous implementation using `mb_substr()` was extremely slow for larger inputs.

### CODE REVIEW

File: lib/Parser/Ntriples.php [hunk 1]
```diff
@@ -39,48 +41,50 @@
...
+        if (stripos($str, '\\u') === false) {
             return $str;
         }
 
-        while (preg_match('/\\\(U)([0-9A-F]{8})/', $str, $matches) ||
-               preg_match('/\\\(u)([0-9A-F]{4})/', $str, $matches)) {
-            $no = hexdec($matches[2]);
+        while (preg_match('/\\\\U([0-9A-F]{8})/', $str, $matches) ||
+        preg_match('/\\\\u([0-9A-F]{4})/', $str, $matches)) {
+            // can be replaced with mb_chr() in PHP >=7.2
```
Comment (k00ni) [inline lib/Parser/Ntriples.php:86]: Can you add a `TODO` in front please?

File: lib/Parser/Ntriples.php [hunk 2]
```diff
@@ -120,51 +124,48 @@ protected function parseNtriplesSubject($sub, $lineNum)
             }
         } else {
             throw new Exception(
-                "Failed to parse subject: $sub",
-                $lineNum
+                    "Failed to parse subject: $sub",
```
Comment (k00ni) [inline lib/Parser/Ntriples.php:127]: Minor: if i see correctly, these lines are indented twice. Can you remove 4 white spaces?

File: lib/Parser/Ntriples.php [hunk 3]
```diff
@@ -120,51 +124,48 @@ protected function parseNtriplesSubject($sub, $lineNum)
...
+                    'value' => $this->unescapeString(substr($matches[3], 1, -1))
                 );
-            }
-        } else {
-            throw new Exception(
-                "Failed to parse object: $obj",
-                $lineNum
-            );
+            default:
+                throw new Exception("Failed to parse object: " . $matches[3], $lineNum);
```
Comment (k00ni) [inline lib/Parser/Ntriples.php:168]: Can `$matches` be empty at this point? Or not having 4 entries?

File: lib/Parser/Ntriples.php [hunk 4]
```diff
@@ -186,7 +187,7 @@ public function parse($graph, $data, $format, $baseUri)
 
         if ($format != 'ntriples') {
             throw new \EasyRdf\Exception(
-                "EasyRdf\\Parser\\Ntriples does not support: $format"
+                    "EasyRdf\\Parser\\Ntriples does not support: $format"
```
Comment (k00ni) [inline lib/Parser/Ntriples.php:190]: ... again, twice indented. Please remove these additional white spaces.

I know its not your code, but would it be possible to use `Ntriples::class` instead of `EasyRdf\\Parser\\Ntriples` here?

File: lib/Parser/Turtle.php [hunk 5]
```diff
@@ -43,7 +43,7 @@
 use EasyRdf\RdfNamespace;
 
 /**
- * Class to parse Turtle with no external dependencies.
+ * Class to parse Turtle with no external dependancies.
```
Comment (k00ni) [inline lib/Parser/Turtle.php:46]: Why do you use the `Archaic form of dependency` here? Just curious.

Ref: https://en.wiktionary.org/wiki/dependancy#English

Comment (zozlak) [reply lib/Parser/Turtle.php:46]: I guess it's simply what was there when I forked the easyrdf :)

File: lib/Parser/Ntriples.php (hunk 2 above)
Comment (zozlak) [reply lib/Parser/Ntriples.php:127]: > There are a few new functions and properties which are used in a static way, e.g. `Ntriples::escapeLiteral`. Is there any reason why these where made `static`?

That's because they are stateless (depend solely on their parameters) and don't generate any side effects. Making them static indicates that clearly and makes it easier to reuse them elsewhere.

File: lib/Parser/Turtle.php (hunk 5 above)
Comment (k00ni) [reply lib/Parser/Turtle.php:46]: Alright.

File: lib/Parser/Ntriples.php (hunk 3 above)
Comment (zozlak) [reply lib/Parser/Ntriples.php:168]: You are right that looking at the regex in line 200 it can't.

On the other hand the `parseNtriplesObject()` can't assure with which regex matches it is called so checking the input may make sense. But following this way of thinking we should also check if `$matches[3]` exists and throw an error if it doesn't.

File: lib/Parser/Ntriples.php (hunk 2 above)
Comment (k00ni) [reply lib/Parser/Ntriples.php:127]: OK, got your point and i'll accept.

Unfortunately you answered on the wrong question.

**Comment by zozlak** [comment]
> There are a few new functions and properties which are used in a static way, e.g. `Ntriples::escapeLiteral`. Is there any reason why these where made `static`?

That's because they are stateless (depend solely on their parameters) and don't generate any side effects. Making them static indicates that clearly and makes it easier to reuse them elsewhere.

File: lib/Parser/Ntriples.php (hunk 3 above)
Comment (k00ni) [reply lib/Parser/Ntriples.php:168]: > On the other hand the parseNtriplesObject() can't assure with which regex matches it is called so checking the input may make sense. But following this way of thinking we should also check if $matches[3] exists and throw an error if it doesn't.

I was thinking the same. Can you add a few checks here? **Or** instead, maybe don't add `$matches[3]` to the exception message or use something like `json_encode` to make it a valid string (regardless of its content). As you like.

File: lib/Parser/Ntriples.php (hunk 2 above)
Comment (k00ni) [reply lib/Parser/Ntriples.php:127]: Can you run `vendor/bin/php-cs-fixer fix` on the console? It was installed via composer and automatically corrects coding-style errors.

Comment (zozlak) [reply lib/Parser/Ntriples.php:127]: `composer update && ls vendor/bin/` gives me `php-coveralls  phpcs  phpunit  yaml-lint` and `grep php-cs-fixer composer.json` returns no results. Should `friendsofphp/php-cs-fixer` be added do development dependencies in the composer.json?

**Comment by zozlak** [comment]
I can see one test fails for PHP 5.5. To be honest I don't think it's worth to track it down as 5.5 is obsolete for ages now and it's usage share is rather neglectable (see e.g. https://blog.packagist.com/php-versions-stats-2019-2-edition/). I would propose to bump the minimum PHP version requirements to 5.6

**Comment by k00ni** [comment]
> I can see one test fails for PHP 5.5. 

Its usually one of the samples which couldn't make a connection to dbpedia.org in time and runs into a timeout. I restarted the whole build and its ran successfully. https://travis-ci.com/github/sweetyrdf/easyrdf/builds/169575432

File: lib/Parser/Ntriples.php (hunk 4 above)
Comment (k00ni) [reply lib/Parser/Ntriples.php:190]: As mentioned, ignore coding styles issues for now.

**Comment by zozlak** [comment]
Sounds reasonable :) I rather won't have time today but I should manage over the weekend.

By the way the Scrutenizer is wrong [here](https://scrutinizer-ci.com/g/sweetyrdf/easyrdf/inspections/453dea3b-07e0-4b56-aea8-f537046c7587/issues/files/lib/Serialiser/Ntriples.php?status=new&orderField=path&order=asc&honorSelectedPaths=0). But as my syntax is rather cryptic - this map can be initialized statically, just I was to lazy to write it for 42 characters - maybe I should rewrite it to a static initialization.

**Comment by zozlak** [comment]
Changes made:

* `\EasyRdf\Serializer\Ntriples::$iriEscapeMap` is now initialized statically for better code readability
    * also the _TODO_ added [here](https://github.com/acdh-oeaw/easyrdf/blob/sweetyrdf/lib/Parser/Ntriples.php#L86)
* `\EasyRdf\Parser\Ntriples::parseNtriplesObject()`:
    * got additional input check
    * got in-line comment describing why it has so ugly API
    * was made private as it depends so heavily on the regex used in `parse()` that I don't consider it safe to expose it as protected

File: lib/Parser/Ntriples.php [hunk 6]
```diff
@@ -131,11 +131,30 @@ protected function parseNtriplesSubject($sub, $lineNum)
...
+     * @param array $matches an array providing the RDF triple object value 
+     *   in the fourth element, datatype in sixth element and lang tag in 
+     *   seventh element
+     * @param type $lineNum 
+     * @return array 
+     * @throws Exception
      * @ignore
      */
-    protected function parseNtriplesObject(&$matches, $lineNum)
+    private function parseNtriplesObject(&$matches, $lineNum)
```
Comment (k00ni) [inline lib/Parser/Ntriples.php]: Why did you change it to private? It's not accessible by child classes anymore. Please revert.

Comment (k00ni) [reply lib/Parser/Ntriples.php]: Sorry missed your comment

> was made private as it depends so heavily on the regex used in parse() that I don't consider it safe to expose it as protected

I found some problems in your `parseNtriplesObject`:

* [ ] You renamed the first parameter and made it call-by-reference. Is that API break really necessary?
* [ ] Reading the description i assume you changed the expected content of that parameter too? `@param array $matches an array providing the RDF triple object value in the fourth element, datatype in sixth element and lang tag in seventh element`

Comment (zozlak) [reply lib/Parser/Ntriples.php]: > Is that API break really necessary?

Depends on how you define necessary. My way of reasoning was as follows:

* This method is called only once in the whole library (both before and after my changes) - in a single line of the `parse()` method.
    * By the way it's the same for `parseNtriplesSubject()`.
* This means the only reason to extract it as a separate method is to provide better readability of the `parse()` method but nothing more.
    * I guess declaring it protected by original author could be meant for reuse in the Turtle parser but as the Turtle parser implementation works in a slightly different way (all in all Turtle parsing is much more complex) this never happened. And in my opinion it will never happen.
* If I know the `parseNtriplesObject()` method is in fact internal, let's declare it private and implement in a performant way (meaning avoiding copying literal value - all other triple elements are typically of neglectable size but a string literal values can be big).
    * Ntriples row is simple enough to be parsed with a one regex which can extract all the important parts (subject, property, object value, object datatype, object language tag). This already happens in `parse()`.
    * Once triple parts are extracted it makes no sense (from a performance point of view) to copy extracted values but to avoid it regex matches array has to be passed by reference and we should keep its original structure (with values `parseNtriplesObject()` is interested in being its 3rd, 6th and 7th element) ending with pretty strange `parseNtriplesObject()` API.

**Comment by k00ni** [comment]
My argument was leaning towards cases in which people created a child class of `Ntriples` and used this function. Following semantic version these functions would be considered part of the API. Its theoretical and may not the case. One could argue in both ways and it would lead to nothing here. 

I agree with your arguments. Thank you for taking the time and pointing it out. That level of detail i was not aware of.

It seems we are done here. Is there anything you wanna do or can i merge it?

**Comment by k00ni** [comment]
Thank you for your effort @zozlak!
